### PR TITLE
Fix performance for next and adapter chains

### PR DESCRIPTION
**Pull request by denis-tingajkin**
## Motivation

https://github.com/networkservicemesh/sdk/issues/297

With this patch per each adapter in chain, adapter will be called once(in master version twice)

### CODE REVIEW

File: pkg/networkservice/common/updatepath/updatepath_test.go [hunk 1]
```diff
@@ -41,7 +41,7 @@ func TestChain(t *testing.T) {
 	want := &networkservice.Connection{
 		Id: "conn-2",
 		Path: &networkservice.Path{
-			Index: 0,
+			Index: 2,
```
Comment (denis-tingajkin) [inline pkg/networkservice/common/updatepath/updatepath_test.go:44]: @edwarnicke I found that in a master version by some reason index did not increment. Can you please verify this line change?

File: pkg/networkservice/core/adapters/server_to_client.go [hunk 2]
```diff
@@ -36,9 +38,27 @@ func NewServerToClient(server networkservice.NetworkServiceServer) networkservic
...
-	return s.server.Request(ctx, in)
+	conn, err := s.server.Request(ctx, in)
+	if err != nil {
+		return nil, err
+	}
+	if in == nil {
+		in = &networkservice.NetworkServiceRequest{}
+	}
+	in.Connection = conn
+	if !next.Done(ctx) {
```
Comment (denis-tingajkin) [inline pkg/networkservice/core/adapters/server_to_client.go:49]: NOTE: I've added flag `done` for pkg next to support case when we have a chain of 
s -> adapted c - > s and for example client did broke the chain, then we should not call next server.

File: pkg/networkservice/common/updatepath/updatepath_test.go (hunk 1 above)
Comment (denis-tingajkin) [reply pkg/networkservice/common/updatepath/updatepath_test.go:44]: As I can see here is used to update servers and looks like here should be 2. Looks like this patch also fixed issue with this test :)

File: pkg/networkservice/common/updatepath/updatepath_test.go [hunk 3]
```diff
@@ -73,5 +73,5 @@ func TestChain(t *testing.T) {
 
 	got, err := server.Request(context.Background(), request)
 	assert.Equal(t, want, got)
-	assert.Equal(t, nil, err)
+	assert.NoError(t, err)
```
Comment (edwarnicke) [inline pkg/networkservice/common/updatepath/updatepath_test.go:76]: Thanks for fixing this :)

**Pull request merged**